meta-sokol-flex-distro/lib/bblayers/flex_utils.py

```diff
@@ -254,50 +254,50 @@ def do_dump_licenses(self, args):
+            appends = self.tinfoil.get_file_appends(fn)
+            data = self.tinfoil.parse_recipe_file(fn, appendlist=appends)
+
+            bpn = data.getVar('BPN')
+            pv = data.getVar('PV')
+            lc = data.getVar('LICENSE')
+
+            if not args.sourceinfo:
+                unique_lines.add('%s,%s,%s\n' % (bpn, pv, lc))
+            else:
+                # unset su, otherwise we get previous value if there is no current
+                su = ''
+                for url in data.getVar('SRC_URI').split():
+                    scheme, host, path, user, passwd, param = bb.fetch.decodeurl(url)
+                    if scheme != 'file':
+                        su = bb.fetch.encodeurl((scheme, host, path, '', '', ''))
+                hp = data.getVar('HOMEPAGE')
+
+                unique_lines.add('%s,%s,%s,%s,%s\n' % (bpn, pv, lc, su, hp))
+
+        sorted_lines = sorted(unique_lines)
         omode = 'a' if args.append else 'w'
+
         with open(filename, omode) as f:
-            for recipe in fetch_recipes:
-                try:
-                    fn = depgraph['pn'][recipe]['filename']
-                except KeyError:
-                    mc = self.tinfoil.config_data.getVar('BBMULTICONFIG')
-                    if not mc:
-                        raise Exception("Could not find key '%s' in depgraph and no multiconfigs defined" % recipe)
-                    for cfg in mc.split():
-                        try:
-                            nkey = f"mc:{cfg}:{recipe}"
-                            fn = depgraph['pn'][nkey]['filename']
-                        except KeyError:
-                            continue
-                if not fn:
-                    raise Exception("Could not find recipe for '%s' in depgraph" % recipe)
-
-                real_fn, cls, mc = bb.cache.virtualfn2realfn(fn)
-                appends = self.tinfoil.get_file_appends(fn)
-                data = self.tinfoil.parse_recipe_file(fn, appendlist=appends)
-
-                pn = data.getVar('PN')
-                pv = data.getVar('PV')
-                lc = data.getVar('LICENSE')
-
-                if not args.sourceinfo:
-                    f.write('%s,%s,%s\n' % (pn, pv, lc))
-                else:
-                    # unset su, otherwise we get previous value if there is no current
-                    su = ''
-                    for url in data.getVar('SRC_URI').split():
-                        scheme, host, path, user, passwd, param = bb.fetch.decodeurl(url)
-                        if scheme != 'file':
-                            su = bb.fetch.encodeurl((scheme, host, path, '', '', ''))
-                    hp = data.getVar('HOMEPAGE')
-
-                    f.write('%s,%s,%s,%s,%s\n' % (pn, pv, lc, su, hp))
-
-        # remove any duplicates added due to append flag
-        uniqlines = set(open(filename).readlines())
-        with open(filename, 'w') as f:
-            f.writelines(uniqlines)
```

Comment:
If you decide to leave the order as is, you can use the `unique_everseen` [recipe](https://docs.python.org/3/library/itertools.html?highlight=unique_everseen#itertools-recipes) in the python itertools recipes in the python docs, otherwise `sorted` will do.

Reply:
> In the case of omode=append and an already existing file, this will essentially randomize the line order of the existing file before we append to it

From what I have tested, the output file remains sorted in both modes, i.e. "w" or "a". However, I do see that in the case of append mode, the final output file is not sorted anymore (since all the lines got **appended** to it), and it contains a lot of duplicates. I guess I will have to perform a read first and then write the consolidated set.
Thanks for the review and for sharing the unique_everseen recipe, I'll take a look at it.